### ISSUE TITLE
How can we know multi GPU is working?

### ISSUE THREAD

**Issue by george-roussos** [kan-bayashi/ParallelWaveGAN#154]
It is mentioned in the paper that using more GPUs accelerates the training. I have three NVIDIA K80s and using the flags 

`--nnodes 1 --nproc_per_node 3 -c`

Binds all three GPUs and ramps them up to 98% usage, however, I cannot see any decrease in waiting time or epoch rounds and leaving it overnight, did not return any marginally better results. Am I doing anythign wrong? How can we know it is actually working? I tried to set `--nnodes 3` but training never even started.

**Comment by kan-bayashi** [comment]
Please use `--n_gpus` option.
E.g. `run.sh --stage 2 --n_gpus 3` 
https://github.com/kan-bayashi/ParallelWaveGAN/blob/d3b2b7b6b54a66a4e62b6533f8f990564258dabf/egs/ljspeech/voc1/run.sh#L121-L125

**Comment by george-roussos** [comment]
Thanks! But I am not using any recipes. So I am trying to run things like

`sudo CUDA_VISIBLE_DEVICES="0,1,2" python3 parallel_wavegan/distributed/launch.py --nproc_per_node 3 -m parallel_wavegan/bin/train.py --config ../config1.yml --resume /home/georgios/s_pwgan/checkpoint-456504steps.pkl --outdir /home/georgios/s_pwgan --train-dumpdir /home/georgios/s_pwgan/wavs --dev-dumpdir /home/georgios/s_pwgan/eval`

So when running nvidia-smi I can see that all thre GPUs are being used fully, but when looking at the training, it doesn't look like the waiting time it gives is decreasing, or step is faster. Is this how it should be? How do I know it is working correctly?

**Comment by kan-bayashi** [comment]
As I wrote in [README](https://github.com/kan-bayashi/ParallelWaveGAN/tree/master/egs#how-to-run-the-recipe), the batch-size will be multiplied by the number of gpus.
If you want to keep the same batch-size, please decrease the batch-size.

**Comment by george-roussos** [comment]
Excuse the noob (me), but I do not get how this works. Right, so if we keep the same batchsize (in my case it is 4), it is multiplied by 3, so that gives a batch size of 12. But if this is the culprit, then how is it that 12 fits in my GPU? Because there is no way it would fit. 

I can try and decrease the batch size to 2, so then it would be 2*3=6. But would that increase performance/decrease wait times?

**Comment by kan-bayashi** [comment]
We use `DistributedDataParallel`. 
https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
In the case of `batch_size=3` and `n_gpus=4`, each gpu process batch_size=3 and then summarized the loss among the GPUs, update parameters.
So if you compared with `batch_size=3` and `n_gpus=1`, the speed of each iteration will decrease.

**Comment by kan-bayashi** [comment]
The advantage of multi-gpus in `DataParallel` is to use the large batch-size for training. 
So if the required batch size is small (it can process within a single gpu), the use of multi-GPUs is not effective.

**Comment by george-roussos** [comment]
Oh wow. Thanks for the explanation.

Well I decreased the batch size to 2 (it was 4).

It seems that the waiting time is cut in half (now I get 250 hours instead of 500), but I am dubious over performance gain, since this batch size seems too small to me. So I guess, my question is what you would recommend, when I have 3 GPUS. A batch size of 4, or a batch size of 2?

Another issue I seem to be having is that the discriminator loss seems to be increasing.

**Comment by kan-bayashi** [comment]
Basically, I have never use multi-gpu for this training since the single gpu is enough to calculate and I do not have a lot of gpus.
And the batch-size in GAN training looks very sensitive and mysterious, the use of a large batch-size does not always give us better results.
See the comment: https://github.com/kan-bayashi/ParallelWaveGAN/issues/143#issuecomment-631297434

**Comment by george-roussos** [comment]
Thanks for all this info 😀

One more thing. Would you recommend a bigger batch_max_step and a smaller batch size, or vice versa?

**Comment by kan-bayashi** [comment]
I think `bigger batch_max_step and a smaller batch size` is better for the first trial but actually I've never investigated about that. So if you have no time to check it, it is better to follow the paper's default.

**Comment by george-roussos** [comment]
How come the LibriTTS model takes less time to converge than LJSpeech? LibriTTS mentions 7 days but LJSpeech notes 21 days.

**Comment by kan-bayashi** [comment]
Which config do you compare?
You should focus on the difference of the config, not dataset.

**Comment by george-roussos** [comment]
The one [here](https://github.com/kan-bayashi/ParallelWaveGAN/blob/master/egs/ljspeech/voc1/conf/parallel_wavegan.v3.yaml) and [here](https://github.com/kan-bayashi/ParallelWaveGAN/blob/master/egs/libritts/voc1/conf/parallel_wavegan.v1.long.yaml)

**Comment by kan-bayashi** [comment]
Discriminater is different.
See the comment in the header.

**Comment by george-roussos** [comment]
I have been trying for the last two days to finetune a vocoder to my single speaker, but I have been having no luck. The loss has just been spiking. I saw that the discriminator type is **ResidualParallelWaveGANDiscriminator**. What is the difference between this and the LibriTTS config, which has no mention of discriminator type?

**Comment by kan-bayashi** [comment]
ResidualParallelWaveGANDiscriminator is deprecated, I could not find a stable training setting (See #27 #61 #123). Please use the other settings.
I do not want to answer all of your minor questions, so at first please understand by yourself by reading the code and checking the other issues.

**Comment by george-roussos** [comment]
Ohhh, maybe that is why. Yeah I understand, sorry :)
